File: core/runtime/types/skill_types.py
from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, TypeAlias

from core.runtime.storage import JSONValue

if TYPE_CHECKING:
    from core.runtime.types.workspace_types import ServiceMutationRequest, ServiceMutationResult
    from core.runtime.types.import_export_types import ExportArtifactRequest, ExportArtifactResult

JSONObject: TypeAlias = dict[str, JSONValue]


@dataclass(frozen=True, slots=True)
class SkillWorkshopSkillSnapshot:
    object_id: str
    revision_id: str
    revision_number: int
    name: str
    description: str
    instruction: str
    style_scope: str
    is_active: bool
    source_kind: str
    donor_kind: str | None
    payload: JSONObject


@dataclass(frozen=True, slots=True)
class SkillWorkshopVersionSnapshot:
    revision_id: str
    revision_number: int
    parent_revision_id: str | None
    name: str
    instruction: str
    style_scope: str
    is_active: bool
    payload: JSONObject


@dataclass(frozen=True, slots=True)
class SkillWorkshopCompareRequest:
    skill_object_id: str
    left_revision_id: str
    right_revision_id: str


@dataclass(frozen=True, slots=True)
class SkillWorkshopComparison:
    skill_object_id: str
    left_revision_id: str
    left_revision_number: int
    right_revision_id: str
    right_revision_number: int
    structured_diff: JSONObject
    rendered_diff: str


@dataclass(frozen=True, slots=True)
class SkillWorkshopUpsertRequest:
    novel_id: str
    actor: str
    source_surface: str
    skill_object_id: str | None = None
    name: str | None = None
    description: str | None = None
    instruction: str | None = None
    style_scope: str | None = None
    is_active: bool | None = None
    base_revision_id: str | None = None
    revision_reason: str | None = None
    source_ref: str | None = None
    import_mapping: JSONObject | None = None
    source_kind: str = "skill_workshop"


@dataclass(frozen=True, slots=True)
class SkillWorkshopImportRequest:
    donor_kind: str
    novel_id: str
    actor: str
    source_surface: str
    donor_payload: JSONObject
    name: str | None = None
    description: str | None = None
    instruction: str | None = None
    style_scope: str = "scene_to_chapter"
    is_active: bool = True
    source_ref: str | None = None


@dataclass(frozen=True, slots=True)
class SkillWorkshopRollbackRequest:
    skill_object_id: str
    target_revision_id: str
    actor: str
    source_surface: str
    revision_reason: str | None = None


@dataclass(frozen=True, slots=True)
class SkillWorkshopMutationResult:
    object_id: str
    revision_id: str
    revision_number: int
    disposition: str
    policy_class: str
    payload: JSONObject


@dataclass(frozen=True, slots=True)
class SkillWorkshopRequest:
    project_id: str
    novel_id: str
    selected_skill_id: str | None = None
    left_revision_id: str | None = None
    right_revision_id: str | None = None


@dataclass(frozen=True, slots=True)
class SkillWorkshopResult:
    project_id: str
    novel_id: str
    skills: tuple[SkillWorkshopSkillSnapshot, ...]
    selected_skill: SkillWorkshopSkillSnapshot | None
    versions: tuple[SkillWorkshopVersionSnapshot, ...]
    comparison: SkillWorkshopComparison | None


@dataclass(frozen=True, slots=True)
class SkillExecutionRequest:
    skill_name: str
    actor: str
    source_surface: str
    mutation_request: ServiceMutationRequest | None = None
    export_request: ExportArtifactRequest | None = None


@dataclass(frozen=True, slots=True)
class SkillExecutionResult:
    skill_name: str
    mutation_result: ServiceMutationResult | None
    export_result: ExportArtifactResult | None

File: packages/shared/src/types/skill.types.ts
import type { JSONObject } from "./json.js";

// ─── Skill Workshop Snapshots ────────────────────────────────────────────────

export interface SkillWorkshopSkillSnapshot {
  object_id: string;
  revision_id: string;
  revision_number: number;
  name: string;
  description: string;
  instruction: string;
  style_scope: string;
  is_active: boolean;
  source_kind: string;
  donor_kind: string | null;
  payload: JSONObject;
}

export interface SkillWorkshopVersionSnapshot {
  revision_id: string;
  revision_number: number;
  parent_revision_id: string | null;
  name: string;
  instruction: string;
  style_scope: string;
  is_active: boolean;
  payload: JSONObject;
}

// ─── Skill Workshop Requests ─────────────────────────────────────────────────

export interface SkillWorkshopRequest {
  project_id: string;
  novel_id?: string;
  selected_skill_id?: string;
  left_revision_id?: string;
  right_revision_id?: string;
}

export interface SkillWorkshopCompareRequest {
  skill_object_id: string;
  left_revision_id: string;
  right_revision_id: string;
}

export interface SkillWorkshopUpsertRequest {
  project_id: string;
  novel_id?: string;
  actor: string;
  source_surface?: string;
  skill_object_id?: string;
  name: string;
  description?: string;
  instruction: string;
  style_scope?: string;
  is_active?: boolean;
  base_revision_id?: string;
  revision_reason?: string;
  source_ref?: string;
  import_mapping?: JSONObject;
  source_kind?: string;
}

export interface SkillWorkshopImportRequest {
  project_id: string;
  novel_id?: string;
  actor: string;
  source_surface?: string;
  donor_kind: string;
  donor_payload: JSONObject;
  name: string;
  description?: string;
  instruction: string;
  style_scope?: string;
  is_active?: boolean;
  source_ref?: string;
}

export interface SkillWorkshopRollbackRequest {
  skill_object_id: string;
  target_revision_id: string;
  actor: string;
  source_surface?: string;
  revision_reason?: string;
}

// ─── Skill Workshop Results ───────────────────────────────────────────────────

export interface SkillWorkshopComparison {
  skill_object_id: string;
  left_revision_id: string;
  left_revision_number: number;
  right_revision_id: string;
  right_revision_number: number;
  structured_diff: JSONObject;
  rendered_diff: string;
}

export interface SkillWorkshopMutationResult {
  object_id: string;
  revision_id: string;
  revision_number: number;
  disposition: string;
  policy_class: string;
  payload: JSONObject;
}

export interface SkillWorkshopResult {
  project_id: string;
  novel_id: string;
  skills: SkillWorkshopSkillSnapshot[];
  selected_skill: SkillWorkshopSkillSnapshot | null;
  versions: SkillWorkshopVersionSnapshot[];
  comparison: SkillWorkshopComparison | null;
}

// ─── Skill Execution ──────────────────────────────────────────────────────────

export interface SkillExecutionRequest {
  skill_name: string;
  actor: string;
  source_surface?: string;
  mutation_request?: unknown;
  export_request?: unknown;
}

export interface SkillExecutionResult {
  skill_name: string;
  mutation_result: unknown;
  export_result: unknown;
}
